src/v/raft/replicate_batcher.cc

```diff
@@ -34,13 +35,28 @@ replicate_batcher::replicate_batcher(consensus* ptr, size_t cache_size)
   , _max_batch_size_sem(cache_size)
   , _max_batch_size(cache_size) {}
 
-ss::future<result<replicate_result>> replicate_batcher::replicate(
+replicate_stages replicate_batcher::replicate(
   std::optional<model::term_id> expected_term, model::record_batch_reader&& r) {
-    return do_cache(expected_term, std::move(r)).then([this](item_ptr i) {
-        return _lock.with([this] { return flush(); }).then([i] {
-            return i->_promise.get_future();
-        });
-    });
+    ss::promise<> enqueued;
+    auto enqueued_f = enqueued.get_future();
+    auto f = do_cache(expected_term, std::move(r))
+               .then_wrapped([this, enqueued = std::move(enqueued)](
+                               ss::future<item_ptr> f) mutable {
+                   try {
+                       auto i = f.get();
+                       enqueued.set_value();
+                       return _lock.with([this] { return flush(); }).then([i] {
+                           return i->_promise.get_future();
+                       });
+                   } catch (...) {
+                       enqueued.set_exception(std::current_exception());
+                       return ss::make_ready_future<result<replicate_result>>(
+                         errc::replicate_batcher_cache_error);
+                   }
+               })
+               .finally([this] { _ptr->_probe.replicate_done(); });
+
+    return replicate_stages(std::move(enqueued_f), std::move(f));
```

Comment:
I can't tell if this needs a `wrap_stages_with_gate`, but I see no gate, so maybe not.

Reply:
It is wrapped in consensus.cc